contracts/src/v1/Calls.sol

```diff
@@ -34,40 +48,203 @@ library Assets {
+        Channel storage ch = Functions.ensureChannel(channelID);
+        return (ch.inboundNonce, ch.outboundNonce);
+    }
 
-        return _sendTokenCosts(destinationChain, destinationChainFee, maxDestinationChainFee);
+    function channelOperatingModeOf(ChannelID channelID) external view returns (OperatingMode) {
+        Channel storage ch = Functions.ensureChannel(channelID);
+        return ch.mode;
+    }
+
+    // @dev Get token address by tokenID
+    function tokenAddressOf(bytes32 tokenID) external view returns (address) {
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+        return $.tokenAddressOf[tokenID];
     }
 
-    function _sendTokenCosts(ParaID destinationChain, uint128 destinationChainFee, uint128 maxDestinationChainFee)
+    /*
+    * Internal functions
+    */
+
+    // Convert foreign currency to native currency (WND/PAS/KSM/DOT -> ETH)
+    function _convertToNative(UD60x18 exchangeRate, UD60x18 multiplier, UD60x18 amount)
+        internal
+        view
+        returns (uint256)
+    {
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+
+        UD60x18 ethDecimals = convert(1e18);
+        UD60x18 foreignDecimals = convert(10).pow(convert(uint256($.foreignTokenDecimals)));
+        UD60x18 nativeAmount =
+            multiplier.mul(amount).mul(exchangeRate).div(foreignDecimals).mul(ethDecimals);
+        return convert(nativeAmount);
+    }
+
+    // Calculate the fee for accepting an outbound message
+    function _calculateFee(Costs memory costs) internal view returns (uint256) {
+        PricingStorage.Layout storage pricing = PricingStorage.layout();
+        UD60x18 amount = convert(pricing.deliveryCost + costs.foreign);
+        return costs.native + _convertToNative(pricing.exchangeRate, pricing.multiplier, amount);
+    }
+
+    /// @dev Outbound message can be disabled globally or on a per-channel basis.
+    function _ensureOutboundMessagingEnabled(Channel storage ch) internal view {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+        if ($.mode != OperatingMode.Normal || ch.mode != OperatingMode.Normal) {
+            revert Disabled();
+        }
+    }
+
+    // Submit an outbound message to Polkadot, after taking fees
+    function _submitOutbound(Ticket memory ticket) internal {
+        ChannelID channelID = ticket.dest.into();
+        Channel storage channel = Functions.ensureChannel(channelID);
+
+        // Ensure outbound messaging is allowed
+        _ensureOutboundMessagingEnabled(channel);
+
+        // Destination fee always in DOT
+        uint256 fee = _calculateFee(ticket.costs);
+
+        // Ensure the user has enough funds for this message to be accepted
+        if (msg.value < fee) {
+            revert FeePaymentToLow();
+        }
+
+        channel.outboundNonce = channel.outboundNonce + 1;
+
+        // Deposit total fee into agent's contract
+        payable(channel.agent).safeNativeTransfer(fee);
```

Comment:
Same here about the fees, we should collect the fees into the gateway contracts asset hub account.